amr-wind/utilities/tagging/QCriterionRefinement.cpp

```diff
@@ -0,0 +1,154 @@
+#include "amr-wind/utilities/tagging/QCriterionRefinement.H"
+#include "amr-wind/CFDSim.H"
+
+#include "AMReX.H"
+#include "AMReX_ParmParse.H"
+
+namespace amr_wind {
+
+QCriterionRefinement::QCriterionRefinement(const CFDSim& sim)
+    : m_sim(sim)
+    , m_qc_value(
+          m_sim.mesh().maxLevel() + 1, std::numeric_limits<amrex::Real>::max())
+{}
+
+void QCriterionRefinement::initialize(const std::string& key)
+{
+    std::string fname = "velocity";
+
+    const auto& repo = m_sim.repo();
+    if (!repo.field_exists(fname)) {
+        amrex::Abort("FieldRefinement: Cannot find field = " + fname);
+    }
+    m_vel = &(m_sim.repo().get_field(fname));
+
+    amrex::Vector<double> qc_value;
+    amrex::ParmParse pp(key);
+
+    pp.queryarr("values", qc_value);
+
+    if (qc_value.size() == 0u)
+        amrex::Abort(
+            "QCriterionRefinement: Must specify at least one of value");
+
+    {
+        size_t fcount = std::min(qc_value.size(), m_qc_value.size());
+        for (size_t i = 0; i < fcount; ++i) m_qc_value[i] = qc_value[i];
+        m_max_lev_field = fcount - 1;
+    }
+}
+
+void QCriterionRefinement::operator()(
+    int level, amrex::TagBoxArray& tags, amrex::Real time, int)
+{
+    const bool tag_field = level <= m_max_lev_field;
```

Comment:
```suggestion
    const bool tag_field = level <= m_max_lev_field;
    if (!tag-field) return;
```

Do an early return here and don't check it inside the MFIter loop. This will also avoid doing a needless fillpatch operation

Reply:
good catch, I fixed it